Initial use of PETSc coloring algorithms

permcody opened the pull request:
This PR allows us to start using the PETSc coloring algorithms. There is more work before this is complete. First, I probably should check error codes from PETSc everywhere. Also, I need to create new files that will demonstrate the need for these algorithms? I suggest we merge this so that I can play with it on the big problems. 

Also, There's a new issue I've identified. It's possible that the coloring breaks down with the use of periodic boundaries. I'll need to investigate if that's indeed an issue, I'll open a ticket once I've determined that there is a problem and that I can reproduce it.

permcody committed:
Tie into PETSc Graph Coloring algorithms

closes #8804

permcody committed:
Updated affected input files

permcody committed:
Manually reflow a few multiline comments

permcody committed:
Handle periodic boundaries properly

This commit uses the topological elemental neighbors functions
to make sure that we see the right neighbors on the opposite
sides of periodic boundaries

refs #8804

permcody committed:
Switch from vector of vector to flat data structure

refs #8804

permcody commented:
Hold up on this for just a bit. I'm refactoring to move the PETSc stuff into PETSc support.

permcody committed:
Refactor - move routines into better locations

permcody commented:
Ok - this should be ready for review now.

permcody committed:
Solving newly found bug due to advanced coloring

The GrainTracker wouldn't remap to any OP that had zero grains on it
because there were no min distances entries in that case. This commit
makes sure that we have a distance entry if an OP is empty (but
not reserved and not matching).

refs #8804

permcody committed:
Remove unused default ctor

refs #8804

moosebuild commented:
Job `modules debug:linux-gnu` on 0b5d84e : invalidated by @permcody

timeouts

fdkong commented on framework/src/utils/PetscSupport.C:
```diff
@@ -733,6 +737,63 @@ setSinglePetscOption(const std::string & name, const std::string & value)
+
+  // Sparse matrix
+  Mat B;
+  MatConvert(A, MATAIJ, MAT_INITIAL_MATRIX, &B);
```
It should be very easy to construct a sparse matrix based on the original data. There is no need to build a temporary dense matrix.

fdkong replied:
And A is always sequential?

permcody replied:
Well I changed my data structure to match. It's a full dense matrix (ones and zeros) in a row major format. Any advice on building that would be welcome here.

permcody replied:
@fdkong - Rather than fill all these entries manually. I think I'll leave the creation of the dense matrix there followed by using a `MatConvert` method with `MAT_INPLACE_MATRIX`. While that still may not be the optimal way to build a matrix. These are very small matrices. It does get rid of the need for building A then B.

fdkong replied:
It is OK for me.

dschwen commented on framework/src/utils/PetscSupport.C:
```diff
@@ -733,6 +737,63 @@ setSinglePetscOption(const std::string & name, const std::string & value)
+  MatColoringApply(mc, &iscoloring);
+
+  PetscInt nn;
+  IS * isis;
```
I thought ISIS was supposed to be destroyed in 30 days?

permcody commented:
Yeah, We have the best people. I'm sure it'll happen really soon.

On Wed, Mar 29, 2017 at 10:29 AM Daniel Schwen <notifications@github.com>
wrote:

> *@dschwen* commented on this pull request.
> ------------------------------
>
> In framework/src/utils/PetscSupport.C
> <https://github.com/idaholab/moose/pull/8805#discussion_r108690910>:
>
> > +  Mat B;
> +  MatConvert(A, MATAIJ, MAT_INITIAL_MATRIX, &B);
> +
> +  MatColoring mc;
> +  ISColoring iscoloring;
> +  MatColoringCreate(B, &mc);
> +  MatColoringSetType(mc, coloring_algorithm);
> +  MatColoringSetMaxColors(mc, static_cast<PetscInt>(colors));
> +
> +  // Petsc normally colors by distance two (neighbors of neighbors), we just want one
> +  MatColoringSetDistance(mc, 1);
> +  MatColoringSetFromOptions(mc);
> +  MatColoringApply(mc, &iscoloring);
> +
> +  PetscInt nn;
> +  IS * isis;
>
> I thought ISIS was supposed to be destroyed in 30 days?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8805#pullrequestreview-29738150>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIGzij2OdayDnHdBvqJt3aUY5J4w5ks5rqmq9gaJpZM4Mo0MU>
> .
>

dschwen commented on modules/phase_field/src/utils/PolycrystalICTools.C:
```diff
@@ -251,28 +256,26 @@ PolycrystalICTools::buildElementalGrainAdjacencyGraph(
       {
-        adjacency_matrix[i][j] = true;
-        adjacency_matrix[j][i] = true;
+        adjacency_matrix(i, j) = true;
```
If I'm not mistaken you template the matrix on `Real`, but assign `bool`s here. You might as well template it on `PETSCScalar`, right? I'm not too bothered by the implicit cast, but the whole point of the underlying type seems to be compatible with PETSc.

permcody replied:
Well, I really wanted to keep PETSc specific stuff out of MOOSE. I probably should assign Real values though at least.

dschwen commented on modules/phase_field/src/utils/PolycrystalICTools.C:
```diff
@@ -297,45 +300,78 @@ PolycrystalICTools::buildNodalGrainAdjacencyGraph(
-  if (!colorGraph(adjacency_matrix, grain_to_op, n_grains, n_ops, 0))
-    mooseError(
-        "Unable to find a valid Grain to op configuration, do you have enough op variables?");
+  // Use a simple backtracking coloring algorithm
```
any reason for keeping this around?

permcody replied:
For now at least. It does a better job of distributing the OPs among the variables for one thing. It works fine for just about any 2D problem. It also works without PETSc (which I know isn't very important). I'll also want to run comparisons between this algorithm and some of the more advanced ones.

dschwen commented on modules/phase_field/src/utils/PolycrystalICTools.C:
```diff
@@ -297,45 +300,78 @@ PolycrystalICTools::buildNodalGrainAdjacencyGraph(
+MooseEnum
+PolycrystalICTools::coloringAlgorithms()
+{
+  return MooseEnum("legacy bt jp power sl lf id greedy", "legacy");
```
why default to an algorithm that _"does not guarantee a valid coloring"_?

permcody replied:
OK, fair enough. I was trying to squeeze this in without changing a ton of input files. Let's put in a followup issue to switch this over soon. We may have to update several phase-field and marmot input files.

permcody replied:
Looks like 14 tests fail if I switch the default. We'll need to look at these tests as some of them are probably bad and will need to be regolded. Marmot will have a few too. I suggest we go with legacy for now.

permcody replied:
Note to self: Remove "sl, lf, and id" - None of these work for distance one coloring.

dschwen commented on modules/phase_field/src/utils/PolycrystalICTools.C:
```diff
@@ -297,45 +300,78 @@ PolycrystalICTools::buildNodalGrainAdjacencyGraph(
+         "algorithm "
+         "which does not guarantee a valid coloring. \"bt\" is a simple backtracking algorithm "
+         "which will produce a valid coloring but has potential exponential run time. The "
+         "remaining algorithms require PETSc but are recommended for larger problems (See "
```
Won't we pretty much _always_ have PETSc available?

permcody replied:
Yeah, I imagine so. In my mind, I'm still thinking about the case where Trilinos could be used even though we don't do this practically anymore.

dschwen commented on modules/phase_field/tests/reconstruction/2phase_reconstruction4.i:
```diff
@@ -45,7 +45,7 @@
     [./ReconVarIC]
       ebsd_reader = ebsd
-      advanced_op_assignment = true
+      coloring_algorithm = bt
```
Do you have any tests for `legacy`?

permcody replied:
Since legacy is the default, it's used in several of the non-graintracker tests. If I switch the default, we'll see this breakage in several places. I don't think we need a test, this is the normal reduced OP stuff we've been doing in marmot for years.

permcody committed:
Addressing comments

permcody commented:
@dschwen - I've addressed some of your comments but don't want to do anything about removing legacy just yet. Let's get this in and we can start looking at tests to regold. This PR is getting plenty big as is.

BTW - Check out what these advanced algorithms do. I'm not ready to just go whole hog on replacing everything just yet. They don't distribute the colors very well. They just give you a valid coloring.

Hand coded BT algorithm that tries to distribute OPs evenly:
![bt_ic](https://cloud.githubusercontent.com/assets/3036960/24620418/e4899744-185a-11e7-9249-ffa4ce6eb5e9.png)

Advanced PETSc algorithm that produces a valid coloring:
![advanced_ic](https://cloud.githubusercontent.com/assets/3036960/24620431/e9a80f1c-185a-11e7-8da4-34dbeae22c55.png)

permcody commented:
Ping!

fdkong commented on modules/phase_field/src/utils/PolycrystalICTools.C:
```diff
@@ -297,45 +300,78 @@ PolycrystalICTools::buildNodalGrainAdjacencyGraph(
+MooseEnum
+PolycrystalICTools::coloringAlgorithms()
+{
+  return MooseEnum("legacy bt jp power greedy", "legacy");
```
Did you try all these options. Not all petsc coloring algorithms work for distance-1 problems.

permcody replied:
Yes, These are all the options that work. I already removed the three that do nothing for distance-1.
